### ISSUE TITLE
Connect to on-premise generic Apache spark cluster with livy

### ISSUE THREAD

**Issue by GaryLiuGTA** [Microsoft/HDInsightVSCodeExtension#24]
It is not an issue actually, but want to get help/support. We have on-premise spark cluster with livy, which can be connected from my laptop using Jupyter Notebook + Sparkmagic. But have no idea how to connect using Spark-Hive-Tools extension even I personally much prefer vscode due to its powerful functionality and rich extensions. The official mainly focus on instruction of connecting to Azure spark cluster, and only provided very limited information about generic server. 

I am attaching my config.json of sparkmagic below. Can you please suggest me how I can use the below information to connect to our spark cluster and submit my pyspark code interactively/in batch? 
```json
{
  "kernel_python_credentials" : {
    "username": "myusername",
    "password": "xxxxxxx",
    "url": "https://mycompany.hadoop.service:8443/gateway/default/livy/v1/"
  },

  "kernel_scala_credentials" : {
    "username": "",
    "password": "",
    "url": "http://localhost:8998",
    "auth": "None"
  },
  "kernel_r_credentials": {
    "username": "",
    "password": "",
    "url": "http://localhost:8998"
  },

  "logging_config": {
    "version": 1,
    "formatters": {
      "magicsFormatter": { 
        "format": "%(asctime)s\t%(levelname)s\t%(message)s",
        "datefmt": ""
      }
    },
    "handlers": {
      "magicsHandler": { 
        "class": "hdijupyterutils.filehandler.MagicsFileHandler",
        "formatter": "magicsFormatter",
        "home_path": "~/.sparkmagic"
      }
    },
    "loggers": {
      "magicsLogger": { 
        "handlers": ["magicsHandler"],
        "level": "DEBUG",
        "propagate": 0
      }
    }
  },

  "wait_for_idle_timeout_seconds": 15,
  "livy_session_startup_timeout_seconds": 60,

  "fatal_error_suggestion": "The code failed because of a fatal error:\n\t{}.\n\nSome things to try:\na) Make sure Spark has enough available resources for Jupyter to create a Spark context.\nb) Contact your Jupyter administrator to make sure the Spark magics library is configured correctly.\nc) Restart the kernel.",

  "ignore_ssl_errors": true,

  "session_configs": {
    "driverMemory": "4G",
	"executorMemory": "4G",
    "executorCores": 2,
	"conf": {
		"spark.yarn.appMasterEnv.PYSPARK_PYTHON": "/python_env/keras_py3.6_env/bin/python",
		"spark.yarn.executorEnv.PYSPARK_PYTHON": "/python_env/keras_py3.6_env/bin/python"
	}
  },

  "use_auto_viz": true,
  "coerce_dataframe": true,
  "max_results_sql": 2500,
  "pyspark_dataframe_encoding": "utf-8",
  
  "heartbeat_refresh_seconds": 30,
  "livy_server_heartbeat_timeout_seconds": 0,
  "heartbeat_retry_seconds": 10,

  "server_extension_default_kernel_name": "pysparkkernel",
  "custom_headers": {"X-Requested-By": "myusername"}, 
  
  "retry_policy": "configurable",
  "retry_seconds_to_sleep_list": [0.2, 0.5, 1, 3, 5],
  "configurable_retry_policy_max_retries": 8
}
```

**Comment by zesluo** [comment]
Hello Gary, thanks for filing the issue with more information. Are you using the 'link a cluster' ->'Generic Livy endpoint'?

**Comment by GaryLiuTelus** [comment]
Yes. I re-installed the extension, and found there were some errors during the installation: 
![image](https://user-images.githubusercontent.com/39339422/118277118-6eaf2c00-b496-11eb-8272-a62dac639b53.png)

and 

![image](https://user-images.githubusercontent.com/39339422/118277133-72db4980-b496-11eb-8cf6-4ae615c51594.png)

Here is the system information from `About`: 

```
Version: 1.56.2 (user setup)
Commit: 054a9295330880ed74ceaedda236253b4f39a335
Date: 2021-05-12T17:13:13.157Z
Electron: 12.0.4
Chrome: 89.0.4389.114
Node.js: 14.16.0
V8: 8.9.255.24-electron.0
OS: Windows_NT x64 10.0.17763```
Any suggestion?

**Comment by GaryLiuTelus** [comment]
I think I have resolved the second error (SSL one), but the first one still persist.

**Comment by zesluo** [comment]
Thanks for your information. We will investigate it.

**Comment by zesluo** [comment]
Hello Gary, I also got two small questions:
1.the Url you input is under this format?
![image](https://user-images.githubusercontent.com/46309533/119307217-edc61080-bc9d-11eb-97c6-9430ce73e57f.png)

2.is there any error message after you link the cluster or after you run pyspark batch? 
3. your cluster is a  SQL Server big data cluster?

**Comment by GaryLiuGTA** [comment]
Sorry for delayed reply. 
1. The URL is like `https://mycompany.hadoop.service:8443/gateway/default/livy/v1/`
2. I tried it with spark interactive, and it showed the following in the output: 
![image](https://user-images.githubusercontent.com/57276664/119687511-453abc80-be15-11eb-9592-6e14de9d3e61.png)
Looks the jupyter was not properly installed, and the dot at the bottom kept extending for ever. But I can run the regular interactive (select a block of code, right click and select `run selection/line in interactive window`. 
3. It is not a SQL Server big data cluster, it is on premise Apache Spark Cluster. 

Thanks!

**Comment by zesluo** [comment]
Thanks for your reply! any error message after you run pyspark **batch**?

**Comment by GaryLiuGTA** [comment]
This is what I got to run batch from right click menu `Spark: Pyspark Batch`
![image](https://user-images.githubusercontent.com/57276664/119855684-1fc6b500-bee0-11eb-9eb0-1cdb62289ccb.png)

When initialized the link to the cluster, I used `Generic Livy Endpoint` -> `my livy endpoint link` -> `Basic` -> `my user name to server` -> `my password` -> `ignore optional cluster name (Enter)`

**Comment by zesluo** [comment]
did you sign in before?

**Comment by GaryLiuGTA** [comment]
Not in this plugin. But it worked fine in Jupyter + SparkMagic extension.

**Comment by zesluo** [comment]
Got it, thanks. Is it possible or convenient for your to sign in?

**Comment by GaryLiuGTA** [comment]
But how can I sign in? In SparkMagic, only thing I need to do is put the credentials in the configuration file (as shown in my first post above), and it helps me to sign in. While in this extension, the only place I was prompted to provide credentials is  `Generic Livy Endpoint` -> `my livy endpoint link` -> `Basic` -> `my user name to server` -> `my password` -> `ignore optional cluster name (Enter)`. but it did not work.

**Comment by zesluo** [comment]
Thanks for your reply. We are still investigating this, will keep you updated when we get more findings. ^_^
